helga_versionone/tests/util.py

```diff
@@ -0,0 +1,171 @@
+            # This must be from a base class,
+            # copy it so it doesn't get mutated in the base
+            cls._mock_patches = copy(cls._mock_patches)
+        else:
+            cls._mock_patches = set()
+
+        for k, v in dct.iteritems():
+            if isinstance(v, Patch):
+                # If it's a patch add it to patches
+                logger.debug('Adding {0} patch to {1}'.format(k, cls))
+                cls._mock_patches.add(k)
+
+        return super(PatchedTestCaseMeta, cls).__init__(name, bases, dct)
+
+
+class PatchedTestCase(unittest.TestCase):
+    """class properties which are mock._patch instances (returned by mock.patch) will become
+       instance variables which are the mock object.
+
+       patches are started in setUp
+
+       patches are stopped when testMethod completes, either normally or through a deferred
+    """
+
+    __metaclass__ = PatchedTestCaseMeta
+
+    def setUp(self):
+        # Before each test, start the patches
+        # metaclass has decorated the test methods to call clearPatches
+        for attr in self._mock_patches:
+            patcher = getattr(self, attr)
+            setattr(self, '_{0}_patcher'.format(attr), patcher)
+            setattr(self, attr, patcher.start())
+
+        super(PatchedTestCase, self).setUp()
+
+    def clearPatches(self):
+        for p in self._mock_patches:
+            getattr(self, '_{0}_patcher'.format(p)).stop()
+
+
+class V1TestCase(PatchedTestCase):
+    """Base class for all helga_versionone tests
+       helga_versionone.get_v1 is always patched to return self.v1 - a MagicMock()
+       helga_versionone.settings is always patched to settings_stub
+       helga_versionone.db is always patched
+    """
+
+    nick = 'me'
+    channel = '#bots'
+
+    db = patch('helga_versionone.db')
+    settings = patch('helga_versionone.settings', settings_stub)
+    get_v1 = patch('helga_versionone.get_v1')
+
+    def setUp(self):
+        # Starts the patches
+        super(V1TestCase, self).setUp()
+
+        self.v1 = MagicMock()
+        # Depends on "get_v1" being in patches above
+        self.get_v1.return_value = self.v1
+
+        # client is mocked, but doesn't have to be patched.
+        self.client = MagicMock()
+
+    def assertAck(self):
+        self.assertIn(self.client.msg.call_args[0][1], ACKS)
+
+    def _test_command(self, cmd, expected=None, to_nick=False):
+        """Test a version command: !v1 <cmd>"""
+
+        d = helga_versionone.versionone_command(
+            self.v1,
+            self.client,
+            self.channel,
+            self.nick,
+            'unused message',
+            'unused cmd',
+            cmd.split(),
+        )
+
+        if expected is None:
+            return d
+
+        if isinstance(d, Deferred):
+            target = self.nick if to_nick else self.channel
+
+            def check(actual):
+                self.assertEqual(call(target, expected), self.client.msg.call_args)
```

Comment:
Yeah, check() is the callback where the testing actually happens, so it's closed on the expected results and the mock object that will be called during the Deferred. It has to be a callback because the code being tested returns a Deferred.